src/coprocessor/codec/chunk/chunk.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME(shirly): remove following later
+#![allow(dead_code)]
+use super::Result;
+use super::column::Column;
+use coprocessor::codec::Datum;
+use std::sync::Arc;
+use tipb::expression::FieldType;
+
+const CHUNK_INITIAL_CAPACITY: usize = 32;
+
+/// Chunk stores multiple rows of data in Apache Arrow format.
+/// See https://arrow.apache.org/docs/memory_layout.html
+/// Values are appended in compact format and can be directly accessed without decoding.
+/// When the chunk is done processing, we can reuse the allocated memory by resetting it.
+pub struct Chunk {
+    columns: Vec<Column>,
+    // num_virtual_rows indicates the number of virtual rows, which have zero columns.
+    // It is used only when this Chunk doesn't hold any data, i.e. "len(columns)==0".
+    num_virtual_rows: usize,
```

Comment:
I think tikv don't need to implement this `virtual row` feature, it's only used in the tidb-server layer.